url9d04094941_petermikitsh_PizzaDeliverySystem_tgz-pJ8-testdrivers_ManagerReportTestDriverJ8/src/ninja/Driver.java

```diff
@@ -0,0 +1,188 @@
+package ninja;
+
+import java.util.ArrayList;
+
+import model.Address;
+import model.Order;
+
+/**
+ * A Driver will deliver Orders that have the same delivery location.
+ * 
+ * @author  Isioma Nnodum	iun4534@rit.edu
+ * 
+ */
+public class Driver extends Thread {
+	
+	private int _driverId;
+
+	/**
+	 * represents the state of this driver
+	 */
+	private boolean _running;
+
+	/**
+	 * The orders this driver is delivering
+	 */
+	private ArrayList<Order> _currentOrders;
+	
+	/**
+	 * The location to which this driver is traveling null if the driver is idle.
+	 */
+	private Address _deliveryLocation;
+	
+	/**
+	 * Indicates the time at which this driver started its journey to the
+	 *  delivery location.
+	 */
+	private long _startTime;
+
+	/**
+	 * Create a new driver
+	 */
+	public Driver(int id) {
+		
+		_driverId = id;
+		_running = false;
+		_currentOrders = new ArrayList<Order>();
+		_deliveryLocation = null;
+		
+	}
+
+	/**
+	 * Stop this driver
+	 */
+	public void deactivate() {
+		_running = false;
+
+	}
+
+	/**
+	 * Get the next list of orders going to the same location
+	 */
+	private void retreive() {
+
+		while ((0 == _currentOrders.size()) && _running) {
+			
+			_currentOrders = Kitchen.driverRetrieve();
+			
+			// Sleep for a bit.
+        	try {
+        		Thread.sleep( Kitchen.KITCHEN_THREAD_WAIT_TIME );
+        	} catch( InterruptedException ex ) {}
+        	
+		}
+		
+		if( Kitchen.PRINT_DIAGNOSTICS ) {
+			
+			System.out.println( Time.formatTime(SystemTime.getTime()) + ": " +
+					"Driver "+ _driverId +": Retrieved "+ _currentOrders.size() +
+					" order(s) for delivery going to:" + _currentOrders.get(0).getDeliveryLocation().getLocation());
+			
+			for (Order order : _currentOrders ) {
+				System.out.println("\t" + order.getOrderId());
+			}
+			
+		}
+		
+	}
+
+	/**
+	 * Deliver those orders.
+	 */
+	private void deliver() {
+
+		// Since the driver is going to a single location it will take as long
+		// as it take to deliver a single item to deliver all of them.
+		_deliveryLocation = _currentOrders.get(0).getDeliveryLocation();
+		long timeToLocation = Time.scaleUp(Time.convertToMilliseconds( _deliveryLocation.getTimeToLocation()) );
+		
+		// Mark the departure time.
+		_startTime = SystemTime.getTime();
+		
+		if( Kitchen.PRINT_DIAGNOSTICS ) {
+			System.out.println( Time.formatTime(SystemTime.getTime()) + ": " +"Driver "+ _driverId +": Enroute to:");
+			System.out.println("\t" + _deliveryLocation.getLocation());
+			System.out.println("\t\t Will be at location in " + _deliveryLocation.getTimeToLocation() / 2 + " minutes");
+			System.out.println("\t\t Will be at location in " + timeToLocation / 2 + " ninja milliseconds");
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 102:107, 129:136.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.